database/src/lib.rs

```diff
@@ -1,520 +1,110 @@
+/// If you connect to multi-DC cluter, you might experience big latencies while working with the DB.
+/// This is due to the fact that ScyllaDB driver tries to connect to any of the nodes in the cluster
+/// disregarding of the location of the DC. This option allows to filter the connection to the DC you need.
+/// Example: "DC1" where DC1 is located in the same region as the application.
+///
+/// `keepalive_interval: Option<u64>`
+/// ScyllaDB keepalive interval
+///
+/// `max_retry: u8`
+/// Max retry count for ScyllaDB if `strict_mode` is `false`
+///
+/// `strict_mode: bool`
+/// Attempts to store data in the database should be infinite to ensure no data is missing.
+/// Disable it to perform a limited write attempts (`max_retry`)
+/// before skipping giving up and moving to the next piece of data
+#[cfg(feature = "scylla_db")]
+pub struct AdditionalDatabaseOptions {
+    pub preferred_dc: Option<String>,
+    pub keepalive_interval: Option<u64>,
+    pub max_retry: u8,
+    pub strict_mode: bool,
 }
 
-impl scylla::retry_policy::RetryPolicy for CustomDBRetryPolicy {
-    /// Called for each new query, starts a session of deciding about retries
-    fn new_session(&self) -> Box<dyn scylla::retry_policy::RetrySession> {
-        Box::new(CustomRetrySession::new(self.max_retry, self.strict_mode))
-    }
-
-    /// Used to clone this RetryPolicy
-    fn clone_boxed(&self) -> Box<dyn scylla::retry_policy::RetryPolicy> {
-        Box::new(CustomDBRetryPolicy::new(self.max_retry, self.strict_mode))
-    }
-}
-
-struct CustomRetrySession {
-    max_retry: u8,
-    max_retry_count: u8,
-    strict_mode: bool,
+/// Additional database options for postgres database(WIP)
+#[cfg(not(feature = "scylla_db"))]
+pub struct AdditionalDatabaseOptions {}
+
+async fn prepare_db_manager<T>(
+    database_url: &str,
+    database_user: Option<&str>,
+    database_password: Option<&str>,
+    database_options: AdditionalDatabaseOptions,
+) -> anyhow::Result<T>
+where
+    T: BaseDbManager + Send + Sync + 'static,
+{
+    Ok(*T::new(
+        database_url,
+        database_user,
+        database_password,
+        database_options,
+    )
+    .await?)
 }
 
-impl CustomRetrySession {
-    pub fn new(max_retry: u8, strict_mode: bool) -> CustomRetrySession {
-        CustomRetrySession {
-            max_retry,
-            max_retry_count: 0,
-            strict_mode,
-        }
-    }
-
-    /// Decide if we should retry the query
-    pub fn should_retry(&mut self) -> bool {
-        // If `strict_mode` always retry
-        if self.strict_mode {
-            return true;
-        };
-        // If `max_retry_count` is more than `max_retry` don't retry
-        if self.max_retry_count > self.max_retry {
-            return false;
-        };
-        // If `max_retry_count` is less than `max_retry` increment `max_retry_count` and retry
-        self.max_retry_count += 1;
-        true
-    }
+pub async fn prepare_read_rpc_db_manager(
```

Comment:
Can this type of function be defined in the `rpc-server` (and in others accordingly)?

Reply:
I would like to leave it here and not import traits with the implementation of the database into services and then hide them under a feature. Let it be in one place.